File: dali/operators/python_function/jax_function.cc
// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "dali/operators/python_function/jax_function.h"
#include <pybind11/stl.h>
#include <memory>
#include <string>
#include <utility>

namespace dali {

DALI_SCHEMA(JaxPythonFunctionImpl)
    .AddArg("function_id", R"code(Id of the python function)code", DALI_INT64)
    .AddOptionalArg("num_outputs", R"code(Number of outputs)code", 1)
    .NumInput(0, 256)
    .OutputFn([](const OpSpec &spec) { return spec.GetArgument<int>("num_outputs"); })
    .AddOptionalArg<std::vector<TensorLayout>>("output_layouts",
                                               R"code(Tensor data layouts for the outputs.)code",
                                               nullptr)
    .NoPrune()
    .Unserializable()
    .MakeInternal();

// TODO(ktokarski) preserve argument shows up in the function docs
// even though it's not configurable and always true
// Either make it non-preservable or get rid of the argument from the docs

// TODO(ktokarski) Include `sharding` param in the docs

// TODO(ktokarski) Mark it as experimental
DALI_SCHEMA(JaxPythonFunction)
    .DocStr(R"code(Runs the `function` callback, passing in the specified arguments as JAX arrays.

The callback can return 0 or more outputs, all of which must be JAX arrays.

All the inputs and outputs must have the same device placement/sharding.
If the inputs/outputs are placed on GPU(s), the JAX and DALI internal streams
will be synchronized, there is no need to synchronize the launched JAX functions with the host.
)code")
    .NumInput(0, 256)
    .AddOptionalArg("num_outputs", R"code(The number of outputs returned by the `function`.

Function can return no output, in that case the `num_outputs` must be set to 0.
If the `num_outputs` is 1 (the default), callback should return a single JAX array,
for `num_outputs` > 1, callback should return a tuple of JAX arrays.
)code",
                    1)
    .AddOptionalArg<std::vector<TensorLayout>>("output_layouts",
                                               R"code(The layouts of returned tensors.

It can be either a list of strings for all of `num_outputs` respective outputs or a single string
to be set to all of the outputs.

Please note, in DALI, the outermost batch extent is implicit, the layout should
take into account only the sample dimensions.

If the argument is not specified, the `function` has the same number of inputs and outputs and
the dimensionality of respective inputs and outputs is preserved, the layout will be propagated
from the input to the output.)code",
                                               nullptr)
    .NoPrune();


DALI_REGISTER_OPERATOR(JaxPythonFunctionImpl, JaxPythonFunctionImpl<CPUBackend>, CPU);
DALI_REGISTER_OPERATOR(JaxPythonFunctionImpl, JaxPythonFunctionImpl<GPUBackend>, GPU);

}  // namespace dali

File: dali/python/nvidia/dali/plugin/jax/_jax_function.py
# Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Optional

import jax
import jax.dlpack
from jax.sharding import NamedSharding, PositionalSharding, Sharding

from nvidia.dali import ops
from nvidia.dali.ops._operators import python_function


def _with_gpu_dl_tensors_as_arrays(callback):

    def inner(stream, *dl_tensors):
        ts = tuple(jax.dlpack.from_dlpack(t) for t in dl_tensors)
        out = callback(*ts)
        if out is not None:
            out = out if isinstance(out, (tuple, list)) else (out,)
            return tuple(jax.dlpack.to_dlpack(t, stream=stream) for t in out)

    return inner


def _with_cpu_dl_tensors_as_arrays(callback):

    def inner(*dl_tensors):
        ts = tuple(jax.dlpack.from_dlpack(t) for t in dl_tensors)
        out = callback(*ts)
        if out is not None:
            out = out if isinstance(out, (tuple, list)) else (out,)
            return tuple(jax.dlpack.to_dlpack(t) for t in out)

    return inner


def _with_sharding(callback, sharding: Sharding):
    if jax.local_device_count() != 1:
        raise NotImplementedError(
            f"Currently, the `jax_python_function` supports only global/multiprocessing sharding. The number of local devices seen by the process must be 1, got {jax.local_device_count()}"
        )

    if not isinstance(sharding, (NamedSharding, PositionalSharding)):
        raise ValueError(
            f"The value passed as `sharding` must be an instance of `NamedSharding` or `PositionalSharding`, got value of a type {type(sharding)}"
        )

    def as_sharded_array(array):
        array_shape = array.shape
        if isinstance(sharding, NamedSharding):
            global_shape = (sharding.mesh.size * array_shape[0], *array_shape[1:])
        else:
            global_shape = (sharding.shape[0] * array_shape[0], *array_shape[1:])
        return jax.make_array_from_single_device_arrays(global_shape, sharding, [array])

    def as_single_device_array(sharded_array):
        local_arrays = [x.data for x in sharded_array.addressable_shards]
        if len(local_arrays) != 1:
            raise ValueError(
                f"The function returned multiple shards, expected exactly one, got {len(local_arrays)}."
            )
        return local_arrays[0]

    def inner(*arrays):
        sharded_arrays = tuple(as_sharded_array(array) for array in arrays)
        out = callback(*sharded_arrays)
        if out is not None:
            out = out if isinstance(out, (tuple, list)) else (out,)
            return tuple(as_single_device_array(t) for t in out)

    return inner


def _jax_callback_wrapper(function, device, sharding):

    assert device in ("cpu", "gpu")

    dl_pack_wrapper = (
        _with_cpu_dl_tensors_as_arrays if device == "cpu" else _with_gpu_dl_tensors_as_arrays
    )
    return dl_pack_wrapper(function if sharding is None else _with_sharding(function, sharding))


class JaxPythonFunction(
    python_function._get_base_impl("JaxPythonFunction", "JaxPythonFunctionImpl")
):
    ops.register_cpu_op("JaxPythonFunction")
    ops.register_gpu_op("JaxPythonFunction")

    def __init__(self, function, device, sharding: Optional[Sharding] = None, **kwargs):
        self._sharding = sharding
        super().__init__(
            function=_jax_callback_wrapper(function, device, sharding),
            device=device,
            **kwargs,
        )
